src/hal/src/sensors_bmi088_spi.c

```diff
@@ -0,0 +1,360 @@
+#define BMI088_SPI_RX_DMA_IRQ          DMA1_Stream3_IRQn
+#define BMI088_SPI_RX_DMA_IRQHandler   DMA1_Stream3_IRQHandler
+#define BMI088_SPI_RX_DMA_CHANNEL      DMA_Channel_0
+#define BMI088_SPI_RX_DMA_FLAG_TCIF    DMA_FLAG_TCIF3
+
+#define BMI088_SPI_TX_DMA_STREAM       DMA1_Stream4
+#define BMI088_SPI_TX_DMA_IRQ          DMA1_Stream4_IRQn
+#define BMI088_SPI_TX_DMA_IRQHandler   DMA1_Stream4_IRQHandler
+#define BMI088_SPI_TX_DMA_CHANNEL      DMA_Channel_0
+#define BMI088_SPI_TX_DMA_FLAG_TCIF    DMA_FLAG_TCIF4
+
+#define ACC_EN_CS() GPIO_ResetBits(BMI088_ACC_GPIO_CS_PORT, BMI088_ACC_GPIO_CS)
+#define ACC_DIS_CS() GPIO_SetBits(BMI088_ACC_GPIO_CS_PORT, BMI088_ACC_GPIO_CS)
+#define GYR_EN_CS() GPIO_ResetBits(BMI088_GYR_GPIO_CS_PORT, BMI088_GYR_GPIO_CS)
+#define GYR_DIS_CS() GPIO_SetBits(BMI088_GYR_GPIO_CS_PORT, BMI088_GYR_GPIO_CS)
+
+static bool isInit;
+
+static char spiSendByte(char byte)
+{
+  /* Loop while DR register in not emplty */
+  while (SPI_I2S_GetFlagStatus(BMI088_SPI, SPI_I2S_FLAG_TXE) == RESET);
+
+  /* Send byte through the SPI peripheral */
+  SPI_I2S_SendData(BMI088_SPI, byte);
+
+  /* Wait to receive a byte */
+  while (SPI_I2S_GetFlagStatus(BMI088_SPI, SPI_I2S_FLAG_RXNE) == RESET);
+
+  /* Return the byte read from the SPI bus */
+  return SPI_I2S_ReceiveData(BMI088_SPI);
+}
+
+static char spiReceiveByte()
+{
+  return spiSendByte(DUMMY_BYTE);
+}
+
+static void spiDMATransaction(uint8_t reg_addr,
+                              uint8_t *reg_data,
+                              uint16_t len)
+{
+  ASSERT(len < SPI_MAX_DMA_TRANSACTION_SIZE);
+
+  // Disable peripheral before setting up for duplex DMA
+  SPI_Cmd(BMI088_SPI, DISABLE);
+
+  // DMA already configured, just need to set memory addresses and read command byte
+  spiTxBuffer[0] = reg_addr;
+  BMI088_SPI_TX_DMA_STREAM->M0AR = (uint32_t)&spiTxBuffer[0];
+  BMI088_SPI_TX_DMA_STREAM->NDTR = len + 1;
+
+  BMI088_SPI_RX_DMA_STREAM->M0AR = (uint32_t)&spiRxBuffer[0];
+  BMI088_SPI_RX_DMA_STREAM->NDTR = len + 1;
+
+  // Enable SPI DMA Interrupts
+  DMA_ITConfig(BMI088_SPI_TX_DMA_STREAM, DMA_IT_TC, ENABLE);
+  DMA_ITConfig(BMI088_SPI_RX_DMA_STREAM, DMA_IT_TC, ENABLE);
+
+  // Clear DMA Flags
+  DMA_ClearFlag(BMI088_SPI_TX_DMA_STREAM, DMA_FLAG_FEIF4|DMA_FLAG_DMEIF4|
+                DMA_FLAG_TEIF4|DMA_FLAG_HTIF4|DMA_FLAG_TCIF4);
+  DMA_ClearFlag(BMI088_SPI_RX_DMA_STREAM, DMA_FLAG_FEIF3|DMA_FLAG_DMEIF3|
+                DMA_FLAG_TEIF3|DMA_FLAG_HTIF3|DMA_FLAG_TCIF3);
+
+  // Enable DMA Streams
+  DMA_Cmd(BMI088_SPI_TX_DMA_STREAM,ENABLE);
+  DMA_Cmd(BMI088_SPI_RX_DMA_STREAM,ENABLE);
+
+  // Enable SPI DMA requests
+  SPI_I2S_DMACmd(BMI088_SPI, SPI_I2S_DMAReq_Tx, ENABLE);
+  SPI_I2S_DMACmd(BMI088_SPI, SPI_I2S_DMAReq_Rx, ENABLE);
+
+  // Enable peripheral to begin the transaction
+  SPI_Cmd(BMI088_SPI, ENABLE);
+
+  // Wait for completion
+  // TODO: Better error handling rather than passing up invalid data
+  xSemaphoreTake(spiTxDMAComplete, portMAX_DELAY);
```

Comment:
When this line is executed the chip just resets. I have no idea why. The spiTxDMAComplete looks valid.

Reply:
It looks like those are initialized after DMA is enabled. Could it be that we can some sort of transactions before the semaphore is fully initialized?

Reply:
I don't think it is a reset but rather a task switch back to itself... It is all very strange and then it is most often memory related.

Reply:
Found it! It was the DMA interrupts that was missing.